webextension/background.js

```diff
@@ -1,3 +1,82 @@
+browser.contextMenus.create({
+  id: "open-in-this-container",
+  title: "Always Open in This Container",
+  contexts: ["all"],
+});
+
+function getUserContextIdFromCookieStore(cookieStore) {
+  const container = cookieStore.replace("firefox-container-", "");
+  if (container !== cookieStore) {
+    return container;
+  }
+  return false;
+}
+
+browser.runtime.onMessage.addListener((request) => {
+  if (request.neverAsk === true) {
+    storageArea.set({neverAsk: true});
+  }
+});
+
+const storagePrefix = "originContainerMap@@_";
+const storageArea = browser.storage.local;
+
+browser.contextMenus.onClicked.addListener((info, tab) => {
+  const pageUrl = new window.URL(info.pageUrl);
+  const userContextId = getUserContextIdFromCookieStore(tab.cookieStoreId);
+  // Mapping ${pageUrl.origin} to ${userContextId}
+  if (userContextId) {
+    const originStoreKey = `${storagePrefix}${pageUrl.origin}`;
+    storageArea.set({[originStoreKey]: userContextId});
+  }
+});
+
+browser.webRequest.onBeforeRequest.addListener((options) => {
+  if (options.frameId !== 0 || options.tabId === -1) {
+    return {};
+  }
+  const pageUrl = new window.URL(options.url);
+  const originStoreKey = `${storagePrefix}${pageUrl.origin}`;
+  return Promise.all([
+    browser.tabs.get(options.tabId),
+    storageArea.get([originStoreKey, "neverAsk"])
+  ]).then(([tab, storage]) => {
+    const userContextId = getUserContextIdFromCookieStore(tab.cookieStoreId);
+    let neverAsk = false;
+    if (!(originStoreKey in storage)) {
+      return;
+    }
+    if ("neverAsk" in storage) {
+      neverAsk = storage.neverAsk;
+    }
+    const mapUserContextId = storage[originStoreKey];
+    if (userContextId === mapUserContextId) {
+      return {};
+    }
+
+    loadOtherPage(options.url, mapUserContextId, neverAsk);
+    // If the user just opened the tab, we can auto close it
+    if (tab.url === "about:newtab" || tab.url === "about:home") {
+      browser.tabs.remove(tab.id);
+    }
+    return {
+      cancel: true,
+    };
+  }).catch((e) => {
+    throw e;
+  });
+},{urls: ["<all_urls>"], types: ["main_frame"]}, ["blocking"]);
+
+
+function loadOtherPage(url, userContextId, neverAsk = false) {
+  const loadPage = browser.extension.getURL("confirm-page.html");
+  // If the user has explicitly checked "Never Ask Again" on the warning page we will send them straight there
+  if (neverAsk) {
+    browser.tabs.create({url, cookieStoreId: `firefox-container-${userContextId}`});
+  } else {
+    browser.tabs.create({url: `${loadPage}?url=${url}`, cookieStoreId: `firefox-container-${userContextId}`});
```

Comment:
Another point would be: what about if the URL is loaded from a window.open() ? This breaks also 'the web' because you will open a tab, instead of a window, and the parent window will not have a reference to the new one.